pkg/bpf/map_linux.go

```diff
@@ -496,20 +496,32 @@ func (m *Map) OpenParallel() (bool, error) {
 // deleted and reopened without any attempt to retain its previous contents.
 // If the map is marked as non-persistent, it will always be recreated.
 //
+// Logs a warning if map properties have changed.
+//
 // Returns whether the map was deleted and recreated, or an optional error.
 func (m *Map) OpenOrCreate() (bool, error) {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
-	return m.openOrCreate(true)
+	return m.openOrCreate(true, true)
+}
+
+// Like OpenOrCrate, but does not log a warning if map properties have changed.
+// This should only be used for maps that can be completely recreated from the
+// agent without needing to dump the previous map from the datapath.
+func (m *Map) OpenOrCreateWithoutWarning() (bool, error) {
```

Comment:
I'm not sure what a good middle ground is for this in order to avoid multiple variations of the constructor. Depending on the specific map, deleting it may have traffic-impacting consequences, whereas for other maps there are no consequences. Depending on those consequences, it would be nice to have behaviour like this (warn when we know this is the cause of datapath impact).

We could always do this instead with maybe a builder pattern or having an optional input configuration struct that can be used to modify the default behaviours, but maybe @ti-mo has some better ideas.

Reply:
For context: my angle is `bpf.Map` is already too big, does too many things, and has poor separation of concerns. We're already in the process of cutting down on its API to make it easier to use, and adding yet another way of opening a map is a step in the opposite direction; it adds yet another code smell. But that's not even my biggest concern.

My biggest gripe with this approach is that it paves over the insanity that is our current CI with a quick & dirty fix, while not making things any safer for the user. When deployed on a live environment, this will still just emit a log line (a tree falls in the forest.. seriously, the agent is so verbose this will not be noticed) and not actually prevent any data loss. If we're serious about preventing connectivity interruptions, we should fatal the agent or fail/degrade the subsystem (Hive should make this possible eventually) before it's about to do anything destructive.

Back to my point of `bpf.Map` doing too many things; map deletion during upgrades is really the subsystem's responsibility to handle. Only the caller knows exactly what needs to happen on up/downgrade. Instead of this proposal, I'd like to see the map deletion behaviour from `objCheck` removed (because _of course_ a function named 'check' removes a map pin..) and pushed out to `Map.CheckAndUpgrade()` (which also doesn't do what it advertises; should be something like `CheckAndUnpin()`) Then, each subsystem can implement something like [DeleteIfUpgradeNeeded](https://github.com/cilium/cilium/blob/a292a757120127dadec2b6f3c54f2954a8bc69a2/pkg/maps/ctmap/ctmap.go#L705) and emit logs (or not) of the correct log level.

Reply:
>  If we're serious about preventing connectivity interruptions, we should fatal the agent or fail/degrade the subsystem (Hive should make this possible eventually) before it's about to do anything destructive.

Broadly I think this is fair, in particular considering the case "map format change causes breakage" case. However, as a counterpoint there is also the case "the user just wants to adjust map sizes". Currently, this approach where we only warn is a middleground that doesn't require us to solve a difficult problem: Migrating CT maps (nice to have for sure, but at least months of work and quite possibly requires upstream kernel changes). We don't have a solution for this today---the only way to change the CT map size is to change it and break connections. Forcing a crash here would prevent such reconfiguration unless we solve the migration problem. The warning behaviour today provides a signal so that if there is packet drop, you can grep for `warning` in the logs and most likely there is exactly just this one message that states that the reason packets are being dropped is due to the map change. This provides a retrospective explanation without restricting the ability of the user to configure things. This can also reassure the user that after they've changed the settings, subsequent restarts should not cause a similar packet loss issue.

Arguably we could do this "fatal on map diff" logic in CI in order to provide a stronger incentive to avoid such breakages during development. Today this logic exists in the CI code instead of the production code, hence why this PR exists. In the end though there are legitimate use cases for wanting to violate this, and there's a significant amount of effort required in order to mitigate the issue that arises. In that case, while we would prefer zero packet loss, it's better to have a short period of packet loss and allow the user to reconfigure the map (which may be mitigating other issues longer term) rather than preventing the user from configuring this altogether.

> map deletion during upgrades is really the subsystem's responsibility to handle

I think that initially the concern was that this is easy to unintentionally ignore and make breaking changes without realizing. By making it as part of the core API, every map must avoid arbitrary breakages unless they opt out. Structurally this is harder to mess up, because this failure will be brought up during development/testing, then the developer is forced to consider this issue and whether it will cause problems. We could certainly move this out into each of the callers that care - primarily ctmap and nat IIRC, maybe some metrics maps too. Then hopefully we don't introduce additional map types that are populated by the datapath and then start to cause breakage in their usage during upgrade later on? Then again, even if we do move it out to the callers, if we need the same behaviour for ctmap and nat then it seems like there should be a layer that provides this protection over the core API in order to DRY...?

Reply:
^ Thinking a bit more, given that the user can't do anything about the warning, maybe `Info` level is more appropriate and only in the case where it drops traffic. For any other map properties that change when we don't expect any sort of traffic impact, it should probably be debug level.